[14.0] Shopfloor: Add no_prefill_qty option v2

TDu opened the pull request:
Supersedes #421 

ref: rau-30

mmequignon committed:
shopfloor search: split product_from_scan

TDu committed:
shopfloor: checkout no split move line

Disable the splitting of move lines when the quantity is not totally
done.
The diff should be properly taken into account during the put in pack.

TDu committed:
shopfloor: checkout select orther line if not in pack

On `select_line` when a product is scanned and it is not in a package,
the other line not in a package are also selected, but their quantity
done is not changed.

TDu committed:
shopfloor: add prefill qty option on menus

Co-authored-by: Matthieu Méquignon <matthieu.mequignon@camptocamp.com>
Co-authored-by: Juan Miguel Sánchez Arce <juan.sanchez@camptocamp.com>
Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

TDu committed:
shopfloor: fix to_float valiator

TDu committed:
shopfloor: checkout add no prefill qty

Co-authored-by: Juan Miguel Sánchez Arce <juan.sanchez@camptocamp.com>
Co-authored-by: Matthieu Méquignon <matthieu.mequignon@camptocamp.com>
Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

TDu committed:
shopfloor: cluster picking add no prefill qty

Co-authored-by: Juan Miguel Sánchez Arce <juan.sanchez@camptocamp.com>
Co-authored-by: Matthieu Méquignon <matthieu.mequignon@camptocamp.com>
Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

TDu committed:
shopfloor: fix location_content_transfer

TDu committed:
shopfloor: zone picking add no prefill qty

Co-authored-by: Juan Miguel Sánchez Arce <juan.sanchez@camptocamp.com>
Co-authored-by: Matthieu Méquignon <matthieu.mequignon@camptocamp.com>
Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

sebalix commented on shopfloor/components/scan_handler_product.py:
```diff
@@ -15,7 +15,11 @@ class ProductHandler(Component):
     def search(self, identifier):
-        return self._search.product_from_scan(identifier)
+        product = self._search.product_from_scan(identifier)
+        if not product:
+            packaging = self._search.packaging_from_scan(identifier)
+            product = packaging.product_id
```
If the goal (not of this PR, but the general idea) is that all scenarios use this scan handler, we do not solve anything by scanning the product+packaging in the same place where some scenarios have a different flow regarding the packaging (like checkout).
We can keep it like this but we need to be aware of that.

sebalix replied:
Answer from @simahawk in private: all good like this. Scan handler is used as a global feature on Shopfloor, it should not be used in the code of the specific scenario (for this, use `x_from_scan` methods)

sebalix commented on shopfloor/data/shopfloor_scenario_data.xml:
```diff
@@ -1,4 +1,4 @@
-<odoo noupdate="1">
+<odoo noupdate="0">
```
See @simahawk 's comment: https://github.com/OCA/wms/pull/421/files#r1011723561

TDu replied:
Fixed

TDu commented:
The commits related to the improvement of the quantity picker are extracted on https://github.com/OCA/wms/pull/485

jbaudoux committed:
shopfloor_mobile: Fix qty widget - UOM at last

When there is a product packaging with qty 1, the product uom was not
displayed as the last packaging decomposition.
This change ensures the product uom is always displayed after the list
of product packagings.

TDu committed:
shopfloor: fix cluster picking noprefill not set

When scanning a location the quantity should be set when the no prefill
quantity option is not enabled.

jbaudoux committed:
shopfloor_mobile: Improve qty widget in readonly

Co-authored-by: Juan Miguel Sánchez Arce <juan.sanchez@camptocamp.com>

Show packagings if relevant

jbaudoux committed:
shopfloor_mobile: Qty picker initial quantity

Co-authored-by: Juan Miguel Sánchez Arce <juan.sanchez@camptocamp.com>

Allow to initialise the quantity done in the quantity picker.
Always show the quantity todo next to the done quantity.

jbaudoux committed:
shopfloor_mobile: selected move line color

Adapt the color of the move line based on the quantity done vs todo.
Use different colors if less, equal or greater.

jbaudoux commented:
- [x] Zone picking: scan product, then scan invalid pack (with content at an invalid location), error raised "You cannot place it here". The scanned quantity is lost and reset to 0
- [x] Cluster picking: all good
- [x] Checkout: partial pick then scan destination pack => line split is missing (like put in pack is doing)
- [x] Checkout: existing pack, I cannot select a package I created just before that is part of the picking
- [x] Checkout: the new pack button is not enabled (this bug was already there)
- [ ] Checkout: when scanning a second product, only the first one is highlighted (it should also be marked as selected)
- [x] Checkout: edit line quantity, the lines without quantities are not shown anymore after confirmation back on the list of lines

TDu committed:
shopfloor: fix zone picking qty done passed

Following the changes on the no prefill quantity option.
The quantity done is passed on the move line record to the frontend, but
not written on the record.
A few places were not updated properly for this.

TDu commented:
The previous commit handles

> Zone picking: scan product, then scan invalid pack (with content at an invalid location), error raised "You cannot place it here". The scanned quantity is lost and reset to 0

And probably many other cases where the quantity being processed was lost

TDu committed:
shopfloor_mobile: checkout fix new package button

This button should not only be available when there is a carrier.

TDu commented:
Squashed and rebased

TDu committed:
shopfloor: fix tests on rebase

JuMiSanAr commented on shopfloor_mobile/static/wms/src/scenario/checkout_states.js:
```diff
@@ -229,14 +229,13 @@ export const checkout_states = function ($instance) {
                     $instance.odoo.call("set_custom_qty", {
                         picking_id: $instance.state.data.picking.id,
-                        selected_line_ids: $instance.selected_line_ids(),
+                        selected_line_ids: this.selectable_line_ids(),
```
@TDu that should be `$instance.selectable_line_ids` shouldn't it?

TDu replied:
Yes, thank you. I added a fixup

jbaudoux reviewed (approved):
I tested and for me it is now good.

sebalix commented:
One fixup to squash (maybe other commits as well? Or all good?)

mmequignon commented:
> One fixup to squash (maybe other commits as well? Or all good?)

squashed the fixup commit.

simahawk commented on shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py:
```diff
@@ -0,0 +1,107 @@
+        two lines with the same product
+
+        """
+        # FIXME - test to be fixed
```
FIXME or not?

TDu replied:
Fixed see this commit https://github.com/OCA/wms/pull/483/commits/70190d92a3908a9bfe277360bda6654447c50e9d

simahawk commented on shopfloor/tests/test_zone_picking_set_line_destination_no_prefill_qty.py:
```diff
@@ -0,0 +1,117 @@
+    def setUp(self):
+        super().setUp()
+        self.service.work.current_picking_type = self.picking2.picking_type_id
+        self.menu.sudo().no_prefill_qty = True
```
this should be done on class setup

sebalix commented on shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py:
```diff
@@ -0,0 +1,107 @@
+            {move_line: origin_qty_done + 1},
+        )
+
+    def test_scan_package_action_scan_product2_to_increment_qty(self):
+        """
+        two lines with the same product
+
+        """
+        # FIXME - test to be fixed
+        # picking = self._create_picking(
+        #     lines=[(self.product_a, 3), (self.product_a, 1)], confirm=False
+        # )
+        # picking.move_lines._action_confirm(merge=False)
+        # picking.action_confirm()
+        # for move in picking.move_lines:
+        #     self._fill_stock_for_moves(move, in_lot=True)
+        # # self._fill_stock_for_moves(picking.move_lines, in_package=False)
+        # picking.action_assign()
+        # move_lines = picking.move_line_ids
+        # response = self.service.dispatch(
+        #     "scan_package_action",
+        #     params={
+        #         "picking_id": picking.id,
+        #         "selected_line_ids": move_lines.ids,
+        #         "barcode": self.product_a.barcode,
+        #     },
+        # )
+        # self._assert_selected_qties(
+        #     response,
+        #     move_lines[0],
+        #     {move_lines[0]: 1, move_lines[1]: 0},
+        # )
```
What about this test? Still relevant? What is the issue?

jbaudoux replied:
should be re-enabled to see what's failing

simahawk commented on shopfloor_mobile/static/wms/src/wms_utils.js:
```diff
@@ -276,13 +278,19 @@ export class WMSUtils {
-            available_packaging: line.product.packaging,
+        // DEPRECATED - To drop in next versions
+        // Use v-bind="utils.wms.move_line_qty_picker_props(...)" instead
+        alert("wms_utils.move_line_qty_picker_options is deprecated.");
```
ALERT? this has to be logged only :thinking:

simahawk replied:
Please fix this :fire:

jbaudoux replied:
changed to console.log

simahawk reviewed (changes requested):
The only blocking point is the on the JS.
That part seems duplicated - partially - from https://github.com/OCA/wms/pull/485

simahawk commented on shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js:
```diff
@@ -1,7 +1,7 @@
  * Copyright 2020 Camptocamp SA (http://www.camptocamp.com)
  * @author Simone Orsi <simahawk@gmail.com>
- * Copyright 2021 Jacques-Etienne Baudoux (BCIM)
+ * Copyright 2021-2022 Jacques-Etienne Baudoux (BCIM)
```
there's no need to change the (C) year...

jbaudoux replied:
was fixed in https://github.com/OCA/wms/pull/485 ...

jbaudoux replied:
fixed here also now

TDu committed:
shopfloor: checkout no-prefill increment only one line

When the no prefill quantity option is set on scanning a barcode only
one line matching the barcode scanned should be incremented.

mmequignon commented:
/ocabot merge

jbaudoux commented:
@JuMiSanAr I spotted some lines in shopfloor_manual_product_transfer_mobile/static/wms/src/scenario/manual_product_transfer.js
that requires utils.wms.move_line_qty_picker_options to be replaced by v-bind=utils.wms.move_line_qty_picker_props
Can you check?

JuMiSanAr committed:
shopfloor_manual_product_transfer_mobile: use v-bind for qty-picker

sebalix commented:
@JuMiSanAr @mmequignon @TDu 2 tests are failing, I can't merge...

sebalix commented:
All green, let's merge this
/ocabot merge minor
